

#ifndef ONBOARD_PLANNER_SPEED_SPEED_OPTIMIZER_CONFIG_DISPATCHER_H_
#define ONBOARD_PLANNER_SPEED_SPEED_OPTIMIZER_CONFIG_DISPATCHER_H_

#include "absl/types/span.h"
#include "plan_common/path_approx.h"
#include "plan_common/math/segment_matcher/segment_matcher_kdtree.h"
#include "modules/cnoa_pnc/planning/proto/speed_finder_params.pb.h"
#include "modules/cnoa_pnc/planning/proto/trajectory_point.pb.h"
#include "modules/cnoa_pnc/planning/proto/vehicle.pb.h"
#include "object_manager/spacetime_trajectory_manager.h"
#include "plan_common/maps/st_boundary_with_decision.h"

namespace st::planning {

std::optional<SpeedFinderParamsProto::SpeedOptimizerParamsProto>
DispatchSpeedOptimizerConfig(
    absl::Span<const StBoundaryWithDecision> st_boundaries_with_decision,
    const SpacetimeTrajectoryManager& traj_mgr, const PathApprox& path_approx,
    const SegmentMatcherKdtree& path_kd_tree, double av_radius,
    double path_step_length, int path_last_index, double current_v,
    const PathPoint& current_path_point,
    const VehicleGeometryParamsProto& vehicle_geometry_params,
    const SpeedFinderParamsProto::SpeedOptimizerParamsProto&
        raw_speed_optimizer_params,
    const SpeedFinderParamsProto::SpeedOptimizerConfigDispatcherParams&
        config_dispatcher_params);

}  // namespace st::planning

#endif  // ONBOARD_PLANNER_SPEED_SPEED_OPTIMIZER_CONFIG_DISPATCHER_H_
